core-primitives/stf-executor/src/executor_tests.rs

```diff
@@ -0,0 +1,317 @@
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+use crate::{
+	error::{Error, Result},
+	executor::*,
+	traits::{StateUpdateProposer, StfExecuteGenericUpdate, StfExecuteTimedGettersBatch},
+};
+use codec::{Decode, Encode};
+use ita_stf::{
+	test_genesis::{endowed_account, test_genesis_setup, ENDOWED_ACC_FUNDS},
+	AccountId, Balance, ShardIdentifier, State, TrustedCall, TrustedGetter, TrustedGetterSigned,
+};
+use itp_ocall_api::EnclaveAttestationOCallApi;
+use itp_stf_state_handler::handle_state::HandleState;
+use itp_test::{
+	builders::parentchain_header_builder::ParentchainHeaderBuilder,
+	mock::{handle_state_mock::HandleStateMock, onchain_mock::OnchainMock},
+};
+use itp_types::H256;
+use sgx_externalities::SgxExternalitiesTrait;
+use sgx_tstd::panic;
+use sp_core::Pair;
+use sp_runtime::app_crypto::sp_core::blake2_256;
+use std::{sync::Arc, time::Duration, vec, vec::Vec};
+
+// FIXME: Create unit tests for update_states, execute_shield_funds, execute_trusted_call, execute_trusted_call_on_stf #554
+
+pub fn propose_state_update_executes_all_calls_given_enough_time() {
+	// given
+	let (stf_executor, ocall_api, state_handler) = stf_executor();
+	let mrenclave = ocall_api.get_mrenclave_of_self().unwrap().m;
+	let (_, shard) = init_state_and_shard_with_state_handler(state_handler.as_ref());
+	let sender = endowed_account();
+	let signed_call = TrustedCall::balance_transfer(
+		sender.public().into(),
+		sender.public().into(),
+		42,
+	)
+	.sign(&sender.clone().into(), 0, &mrenclave, &shard);
+	let call_operation_hash: H256 =
+		blake2_256(&signed_call.clone().into_trusted_operation(true).encode()).into();
+	let signed_call_two = TrustedCall::balance_transfer(
+		sender.public().into(),
+		sender.public().into(),
+		100,
+	)
+	.sign(&sender.clone().into(), 1, &mrenclave, &shard);
+	let call_operation_hash_two: H256 =
+		blake2_256(&signed_call_two.clone().into_trusted_operation(true).encode()).into();
+
+	let old_state_hash = state_hash(&state_handler.load_initialized(&shard).unwrap());
+
+	// when
+	let batch_execution_result = stf_executor
+		.propose_state_update(
+			&vec![signed_call.clone(), signed_call_two.clone()],
+			&ParentchainHeaderBuilder::default().build(),
+			&shard,
+			Duration::from_secs(1000),
+			|state| state,
+		)
+		.unwrap();
+
+	// then
+	assert_eq!(old_state_hash, batch_execution_result.state_hash_before_execution);
+	assert_eq!(batch_execution_result.executed_operations.len(), 2);
+	assert_eq!(
+		batch_execution_result.get_executed_operation_hashes(),
+		vec![call_operation_hash, call_operation_hash_two]
+	);
+	// Ensure that state has been updated and not actually written.
+	assert_ne!(
+		state_handler.load_initialized(&shard).unwrap(),
+		batch_execution_result.state_after_execution
+	);
+}
+
+pub fn propose_state_update_executes_only_one_trusted_call_given_not_enough_time() {
```

Comment:
This is interesting to know - we always execute at least one call, even for time duration zero - because we check for the remaining time only after the first execution?

Reply:
Yes, that's how it's currently implemented (and hence I tested it accordingly)

Reply:
What specific check do you expect here? Yes it is implemented this way now, that is confirmed by this test.

Reply:
Haha, yes I was very unclear. I think it is illogical that we execute at least one getter even if we don't have any time. We should check in the beginning, if we have any time and return immediately if there is no time left.

Reply:
Okay, I'll create an issue then, if that is okay. I don't think we want to add more changes to this PR? Since that behavior was put in long before this PR, it's just that @haerdib now added tests to confirm the existing behavior.

Reply:
Issue: #568

Reply:
Yes, I totally agree.